url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/spec/args/VarArgList.java

```diff
@@ -0,0 +1,139 @@
+ */
+
+package com.sdicons.scripty.spec.args;
+
+import com.sdicons.scripty.parser.IContext;
+import com.sdicons.scripty.parser.Pair;
+
+/**
+ * There are 3 types of parameters in a variable argument list.
+ * 1. Fixed (always required) arguments, each argument has its own type.
+ * 2. Variable length. These all have the same type. These cannot be pairs.
+ *    It is the same type that can occur min to max times. A repetition of the same type.
+ *    Eg. a list of 1 or more integers.
+ * 3. Named (optional or required). These are pairs at the end of the command line.
+ *    The named parameters can have a default value.
+ *
+ * Warning: The fixed args will be put first in the argument list.
+ *          After that, the (!) named args will be put in the list.
+ *          Finally, the variable part.
+ */
+public class VarArgList
+implements IArgList
+{
+    private FixedArg req[];
+    private VarArg var;
+    private NamedArg[] named;
+    int min, max;
+
+    public VarArgList(FixedArg aReq[], VarArg aVar, int aMin, int aMax, NamedArg[] aNamed)
+    {
+        req = aReq;
+        var = aVar;
+        named = aNamed;
+        min = aMin;
+        max = aMax;
+    }
+
+    /**
+     * The args are expected to have the form ("CMD", arg-1, arg-2, ... arg-n).
+     * The result is in this order:
+     *  - First the fixed.
+     *  - Then !!! the named.
+     *  - Lastly the varargs.
+     */
+    public Object[] guard(Object[] aArgs, IContext aCtx)
+    throws ArgSpecException
+    {
+        // We look for all pairs at the end of the argument list. We will only
+        // consider these trailing pairs.
+        int lStartNamed = aArgs.length - 1;
+        while(lStartNamed > 0 && (aArgs[lStartNamed] instanceof Pair)) lStartNamed --;
+        int lNrVar = lStartNamed - req.length ;
+        if(lNrVar < 0) lNrVar = 0;
+
+        if(min >= 0 && lNrVar < min)
+            throw new ArgSpecException(String.format("Too few arguments of type '%s'. Expected at least %d and received %d.", var.getSpecName(), min, lNrVar));
+        if(max >= 0 && lNrVar > max)
+            throw new ArgSpecException(String.format("Too many arguments of type '%s'. Expected at most %d and received %d.", var.getSpecName(), max, lNrVar));
+
+        // Create a new argument list where we will accumulate the
+        // converted results.
+        Object[] lNewArgs = new Object[1 + req.length + lNrVar + named.length];
+        // Copy the command name to the new argument list, the structure will remain the same.
+        lNewArgs[0] = aArgs[0];
+
+        // Check the fixed.
+        if(aArgs.length - 1 < req.length)
+            throw new ArgSpecException(String.format("Too few arguments. Expected at least %d arguments but received %d.", req.length, aArgs.length - 1));
+
+        // We skip the command name.
+        int lArgIdx = 1;
+        for(int i = 0; i < req.length; i++)
+        {
+            lNewArgs[lArgIdx] = req[i].guard(aArgs, lArgIdx++, aCtx);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;aCtx&#39; (index: 1 | type: IContext). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;aArgs&#39;. The same parameter type is validated here for example: url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java:60. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).